apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/JVMUtil.java

```diff
@@ -0,0 +1,190 @@
+
+    private static Map<String, String> JAR_PATH_MAP = new HashMap<>();
+
+    private static Set<String> JAR_FILE_LIST = new HashSet<>();
+
+    public static List<String> getJarFileNameList() {
+        List<String> jarFileNameList = new ArrayList<>();
+        for (String jarFile : JAR_FILE_LIST) {
+            String jarFileName = jarFile.substring(jarFile.lastIndexOf("/") + 1);
+            jarFileNameList.add(jarFileName);
+        }
+        return jarFileNameList;
+    }
+
+    public static void loadJvmInfo() {
+        loadJarPath();
+        loadJarFileList();
+    }
+
+    private static void loadJarPath() {
+        Instrumentation instrumentation = ServiceManager.INSTRUMENTATION;
+
+        Class[] clzzs = instrumentation.getAllLoadedClasses();
+        Set<ClassLoader> classLoaders = new HashSet<>();
+        for (final Class clzz : clzzs) {
+            ClassLoader classLoader = clzz.getClassLoader();
+            if (classLoader != null) {
+                classLoaders.add(clzz.getClassLoader());
+            }
+        }
+
+        for (ClassLoader classLoader : classLoaders) {
+            classResCalc(classLoader.getClass());
+        }
+    }
+
+    private static void classResCalc(Class c) {
+        try {
+            java.net.URL url = c.getResource("/");
+            if (null != url) {
+                JAR_PATH_MAP.put(url.toString(), "");
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Error msg: {}", e.getMessage());
+        }
+    }
+
+    private static void loadJarFileList() {
+        String[] classPathJar = getPathSeparator().split(OSUtil.getPathSeparator());
+        for (String s : classPathJar) {
+            if (!s.endsWith(".jar")) {
+                continue;
+            }
+            String path = new File(s).getAbsolutePath();
+            JAR_FILE_LIST.add(path);
+        }
+        for (String s : JAR_PATH_MAP.keySet()) {
+            s = s.replace("file:", "");
+            if (s.endsWith("WEB-INF/classes/")) {
+                s = s.replace("WEB-INF/classes/", "WEB-INF/lib/");
+            }
+            File tempDir = new File(s);
+            if (!tempDir.exists() || !tempDir.isDirectory()) {
+                continue;
+            }
+            for (File file : tempDir.listFiles()) {
+                if (!file.getName().endsWith(".jar")) {
+                    continue;
+                }
+                JAR_FILE_LIST.add(file.getAbsolutePath());
+            }
+        }
+        loadSpringBootLib();
+    }
+
+    private static void loadSpringBootLib() {
+        List<String> springBootJars = new ArrayList<>();
+        for (String s : JAR_FILE_LIST) {
+            JarFile jar = null;
+            try {
+                jar = new JarFile(s);
+                if (jar.getManifest() == null || !"org.springframework.boot.loader.JarLauncher".equals(jar.getManifest().getMainAttributes().getValue("Main-Class"))) {
```

Comment:
Can you elaborate more on this?

Reply:
In order to start with a fat jar, spring boot extends the JarFile URL protocol of JDK, customizes its own classloader and jar file protocol handler, and implements the loading mode of jar in jar and jar in directory.

![image](https://user-images.githubusercontent.com/28091237/120884264-b1d36980-c614-11eb-99a4-265afc6d231c.png)

![image](https://user-images.githubusercontent.com/28091237/120884271-ca438400-c614-11eb-822b-940444e7a8f7.png)